discovery/eureka/eureka.go

```diff
@@ -0,0 +1,207 @@
+func (c *SDConfig) NewDiscoverer(opts discovery.DiscovererOptions) (discovery.Discoverer, error) {
+	return NewDiscovery(c, opts.Logger)
+}
+
+// SetDirectory joins any relative file paths with dir.
+func (c *SDConfig) SetDirectory(dir string) {
+	c.HTTPClientConfig.SetDirectory(dir)
+}
+
+// UnmarshalYAML implements the yaml.Unmarshaler interface.
+func (c *SDConfig) UnmarshalYAML(unmarshal func(interface{}) error) error {
+	*c = DefaultSDConfig
+	type plain SDConfig
+	err := unmarshal((*plain)(c))
+	if err != nil {
+		return err
+	}
+	if len(c.Server) == 0 {
+		return errors.New("eureka_sd: empty or null eureka server")
+	}
+	url, err := url.Parse(c.Server)
+	if err != nil {
+		return err
+	}
+	if len(url.Scheme) == 0 || len(url.Host) == 0 {
+		return errors.New("eureka_sd: invalid eureka server URL")
+	}
+	return c.HTTPClientConfig.Validate()
+}
+
+// Discovery provides service discovery based on a Eureka instance.
+type Discovery struct {
+	*refresh.Discovery
+	client *http.Client
+	server string
+}
+
+// New creates a new Eureka discovery for the given role.
+func NewDiscovery(conf *SDConfig, logger log.Logger) (*Discovery, error) {
+	rt, err := config.NewRoundTripperFromConfig(conf.HTTPClientConfig, "eureka_sd", false, false)
+	if err != nil {
+		return nil, err
+	}
+
+	d := &Discovery{
+		client: &http.Client{Transport: rt},
+		server: conf.Server,
+	}
+	d.Discovery = refresh.NewDiscovery(
+		logger,
+		"eureka",
+		time.Duration(conf.RefreshInterval),
+		d.refresh,
+	)
+	return d, nil
+}
+
+func (d *Discovery) refresh(ctx context.Context) ([]*targetgroup.Group, error) {
+	apps, err := fetchApps(ctx, d.server, d.client)
+	if err != nil {
+		return nil, err
+	}
+
+	tg := &targetgroup.Group{
+		Source: "eureka",
+	}
+
+	for _, app := range apps.Applications {
+		targets := targetsForApp(&app)
+		tg.Targets = append(tg.Targets, targets...)
+	}
+	return []*targetgroup.Group{tg}, nil
+}
+
+func targetsForApp(app *Application) []model.LabelSet {
+	targets := make([]model.LabelSet, 0, len(app.Instances))
+
+	// Gather info about the app's 'instances'. Each instance is considered a task.
+	for _, t := range app.Instances {
+		targetAddress := net.JoinHostPort(t.HostName, strconv.Itoa(t.Port.Port))
+		target := model.LabelSet{
+			model.AddressLabel:  model.LabelValue(targetAddress),
+			model.InstanceLabel: model.LabelValue(t.InstanceID),
+
+			appNameLabel:                      lv(app.Name),
+			appInstanceHostNameLabel:          lv(t.HostName),
+			appInstanceHomePageURLLabel:       lv(t.HomePageURL),
+			appInstanceStatusPageURLLabel:     lv(t.StatusPageURL),
+			appInstanceHealthCheckURLLabel:    lv(t.HealthCheckURL),
+			appInstanceIPAddrLabel:            lv(t.IPAddr),
+			appInstanceVipAddressLabel:        lv(t.VipAddress),
+			appInstanceSecureVipAddressLabel:  lv(t.SecureVipAddress),
+			appInstanceStatusLabel:            lv(t.Status),
+			appInstancePortLabel:              lv(strconv.Itoa(t.Port.Port)),
```

Comment:
Port and SecurePort are pointers, what happens if they're nil?

Reply:
I'm going to add a nil check.